File: Numerical_method/Python/lab6/seidel.py
# Gauss-Seidel method

def gauss_seidel(a, b, tol=1e-4):
    n = len(b)
    x = [0.0] * n
    
    while True:
        x_old = x[:]
        converged = True
        
        for i in range(n):
            sum1 = 0.0
            for j in range(i):
                sum1 += a[i][j] * x[j]
            sum2 = 0.0
            for j in range(i + 1, n):
                sum2 += a[i][j] * x[j]
            x[i] = (b[i] - sum1 - sum2) / a[i][i]
            
            if abs((x[i] - x_old[i]) / x[i]) > tol:
                converged = False
        
        if converged:
            break
    
    return x

def main():
    n = int(input("Enter the number of equations: "))
    
    a = [[0.0] * n for _ in range(n)]
    b = [0.0] * n
    
    for i in range(n):
        for j in range(n):
            a[i][j] = float(input(f"Enter the a[{i+1}][{j+1}] element: "))
        b[i] = float(input(f"Enter the b[{i+1}] value: "))
    
    x = gauss_seidel(a, b)
    
    print("The required values are:")
    for i in range(n):
        print(f"x[{i+1}] = {x[i]:.3f}")

if __name__ == "__main__":
    main()

File: Numerical_method/C++/lab6/seidel.cpp
// Gauss-Seidel method

#include <stdio.h>
#include <math.h>
#define E 0.0001
int main()
{
    float x[10], a[10][10], b[10], dummy, sum;
    int n, key = 0, i, j;
    printf("Enter the number of equations: ");
    scanf("%d", &n);
    for (i = 1; i <= n; i++)
    {
        for (j = 1; j <= n; j++)
        {
            printf("Enter the a[%d][%d] element: ", i, j);
            scanf("%f", &a[i][j]);
        }
        printf("Enter the b[%d] value: ", i);
        scanf("%f", &b[i]);
    }
    for (i = 1; i <= n; i++)
    {
        x[i] = b[i] / a[i][i];
    }

label:
    key = 0;
    for (i = 1; i <= n; i++)
    {
        sum = b[i];
        for (j = 1; j <= n; j++)
        {
            if (j != i)
            {
                sum = sum - a[i][j] * x[j];
            }
        }
        dummy = sum / a[i][i];
        if (key == 0)
        {
            if (fabs((dummy - x[i]) / dummy) > E)
            {
                key = 1;
            }
        }
        x[i] = dummy;
    }
    if (key == 1)
    {
        goto label;
    }

    printf("The required value is:\n");
    for (i = 1; i <= n; i++)
    {
        printf("x[%d]=%0.3f", i, x[i]);
        printf("\n");
    }
    return 0;
}
